File: contracts/api/clients/jakarta/src/main/java/app/epistola/client/jakarta/validation/schema/TemplateSchemaValidator.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.jakarta.validation.schema;

import app.epistola.client.jakarta.EpistolaJson;
import app.epistola.client.jakarta.api.TemplatesApi;
import app.epistola.client.jakarta.model.TemplateDto;
import com.networknt.schema.InputFormat;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersion;
import com.networknt.schema.ValidationMessage;
import jakarta.json.Json;
import jakarta.json.JsonObject;
import jakarta.json.JsonReader;
import jakarta.json.JsonValue;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Validates generation data against the JSON Schema declared on the template, before the request
 * leaves the JVM.
 *
 * <p>The template is fetched from the server on first use and the compiled schema is cached.
 *
 * <pre>{@code
 * TemplateSchemaValidator validator = new TemplateSchemaValidator(templatesApi);
 * validator.validate("my-tenant", "default", "monthly-invoice", data);
 * }</pre>
 *
 * <p><strong>Requires an optional dependency.</strong> {@code com.networknt:json-schema-validator}
 * is {@code compileOnly} in this client so the container is not handed a JSON Schema engine (and a
 * Jackson) it did not ask for. Add it to your own build to use this class.
 */
public class TemplateSchemaValidator {

    private static final SpecVersion.VersionFlag DEFAULT_VERSION = SpecVersion.VersionFlag.V202012;

    private final TemplatesApi templatesApi;
    private final SchemaCache cache;

    public TemplateSchemaValidator(TemplatesApi templatesApi) {
        this(templatesApi, new TtlSchemaCache());
    }

    /**
     * @param templatesApi the generated API used to fetch template metadata
     * @param cache        where compiled schemas are kept between calls
     */
    public TemplateSchemaValidator(TemplatesApi templatesApi, SchemaCache cache) {
        this.templatesApi = templatesApi;
        this.cache = cache;
    }

    /**
     * Validates {@code data} against the template's schema.
     *
     * @throws TemplateDataValidationException when the data does not satisfy the schema
     * @throws app.epistola.client.jakarta.api.ApiException when the template cannot be fetched
     */
    public void validate(String tenantId, String catalogId, String templateId, Object data) {
        JsonSchema schema = cache.getOrLoad(tenantId, templateId, () -> loadSchema(tenantId, catalogId, templateId));
        if (schema == null) {
            // No schema on the template — nothing to validate against.
            return;
        }

        Set<ValidationMessage> messages =
                schema.validate(EpistolaJson.jsonb().toJson(data), InputFormat.JSON);
        if (messages.isEmpty()) {
            return;
        }

        List<TemplateDataValidationException.ValidationError> errors = new ArrayList<>(messages.size());
        for (ValidationMessage message : messages) {
            errors.add(new TemplateDataValidationException.ValidationError(
                    message.getInstanceLocation().toString(), message.getMessage(), message.getType()));
        }
        throw new TemplateDataValidationException(errors);
    }

    private JsonSchema loadSchema(String tenantId, String catalogId, String templateId) {
        TemplateDto template = templatesApi.getTemplate(tenantId, catalogId, templateId);
        if (template.getSchema() == null) {
            return null;
        }
        String schemaJson = EpistolaJson.jsonb().toJson(template.getSchema());
        return JsonSchemaFactory.getInstance(detectVersion(schemaJson)).getSchema(schemaJson, InputFormat.JSON);
    }

    /**
     * Reads the schema's own {@code $schema} declaration so a draft-07 template is not validated
     * under 2020-12 rules. Falls back to 2020-12, which is what the editor emits.
     */
    private static SpecVersion.VersionFlag detectVersion(String schemaJson) {
        String declared;
        try (JsonReader reader = Json.createReader(new StringReader(schemaJson))) {
            JsonValue root = reader.readValue();
            if (root.getValueType() != JsonValue.ValueType.OBJECT) {
                return DEFAULT_VERSION;
            }
            JsonObject object = root.asJsonObject();
            JsonValue schemaUri = object.get("$schema");
            if (schemaUri == null || schemaUri.getValueType() != JsonValue.ValueType.STRING) {
                return DEFAULT_VERSION;
            }
            declared = object.getString("$schema");
        } catch (RuntimeException e) {
            return DEFAULT_VERSION;
        }

        if (declared.contains("draft-04")) {
            return SpecVersion.VersionFlag.V4;
        }
        if (declared.contains("draft-06")) {
            return SpecVersion.VersionFlag.V6;
        }
        if (declared.contains("draft-07")) {
            return SpecVersion.VersionFlag.V7;
        }
        if (declared.contains("2019-09")) {
            return SpecVersion.VersionFlag.V201909;
        }
        return DEFAULT_VERSION;
    }
}

File: client-kotlin-spring-restclient/client/src/main/kotlin/app/epistola/client/validation/schema/TemplateSchemaValidator.kt
package app.epistola.client.validation.schema

import app.epistola.client.api.TemplatesApi
import app.epistola.client.infrastructure.Serializer
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.networknt.schema.JsonSchema
import com.networknt.schema.JsonSchemaFactory
import com.networknt.schema.SpecVersion

/**
 * Validates template data against the JSON Schema defined on the template.
 *
 * Fetches the template from the server on first use and caches the compiled schema.
 *
 * ```kotlin
 * val validator = TemplateSchemaValidator(templatesApi)
 * validator.validate("my-tenant", "my-template", myDataMap)
 * ```
 *
 * @param templatesApi The generated [TemplatesApi] used to fetch template metadata.
 * @param cache Schema cache implementation. Defaults to [TtlSchemaCache] with 5-minute TTL.
 * @param objectMapper ObjectMapper for converting data to [JsonNode]. Defaults to the client's shared mapper.
 */
class TemplateSchemaValidator(
    private val templatesApi: TemplatesApi,
    private val cache: SchemaCache = TtlSchemaCache(),
    private val objectMapper: ObjectMapper = Serializer.jacksonObjectMapper,
) {
    /**
     * Validates [data] against the schema of the specified template.
     *
     * @param tenantId Tenant identifier.
     * @param templateId Template identifier.
     * @param data The data object (typically a `Map<String, Any?>`) to validate.
     * @throws TemplateDataValidationException if validation fails.
     * @throws org.springframework.web.client.RestClientResponseException if template fetch fails.
     */
    fun validate(tenantId: String, templateId: String, data: Any) {
        val schema = cache.getOrLoad(tenantId, templateId) { loadSchema(tenantId, templateId) }
            ?: return // No schema defined on template -- nothing to validate

        val dataNode: JsonNode = objectMapper.valueToTree(data)
        val messages = schema.validate(dataNode)

        if (messages.isNotEmpty()) {
            val errors = messages.map { msg ->
                TemplateDataValidationException.ValidationError(
                    path = msg.instanceLocation.toString(),
                    message = msg.message,
                    keyword = msg.type,
                )
            }
            throw TemplateDataValidationException(errors)
        }
    }

    private fun loadSchema(tenantId: String, templateId: String): JsonSchema? {
        val template = templatesApi.getTemplate(tenantId, templateId)
        val schemaObj = template.schema ?: return null
        val schemaNode: JsonNode = objectMapper.valueToTree(schemaObj)
        val versionFlag = detectVersion(schemaNode)
        val factory = JsonSchemaFactory.getInstance(versionFlag)
        return factory.getSchema(schemaNode)
    }

    private fun detectVersion(schemaNode: JsonNode): SpecVersion.VersionFlag {
        val schemaUri = schemaNode.path("\$schema").asText(null) ?: return DEFAULT_VERSION
        return when {
            schemaUri.contains("draft-04") -> SpecVersion.VersionFlag.V4
            schemaUri.contains("draft-06") -> SpecVersion.VersionFlag.V6
            schemaUri.contains("draft-07") -> SpecVersion.VersionFlag.V7
            schemaUri.contains("2019-09") -> SpecVersion.VersionFlag.V201909
            schemaUri.contains("2020-12") -> SpecVersion.VersionFlag.V202012
            else -> DEFAULT_VERSION
        }
    }

    private companion object {
        val DEFAULT_VERSION = SpecVersion.VersionFlag.V202012
    }
}
